[RDY] Newsletter #11: Neovim 0.5 (#235)
neovim/neovim.github.io: _posts/2021-07-05-newsletter.md

To learn more about LSP and how to use it in Neovim, visit [Nvim-lspconfig](https://github.com/neovim/nvim-lspconfig) and read [`:h lsp`](https://neovim.io/doc/user/lsp.html).

 Expect more work on LSP during the 0.5.x development cycle to provide improved configuration options and better coverage of the latest LSP specification (version 3.16 at the point of writing), including semantic highlighting.

## Tree-sitter

Neovim 0.5 adds **experimental** support for tree-sitter, a library that parses a piece of code into a syntax tree in an incremental and error-resilient way; this means that reparsing that code after an edit is very fast, and parsing error due to, e.g., typos remain localized and do not break parsing further down. This tree can then be efficiently queried to obtain syntax information about the code. This allows for improved and/or faster

- syntax highlighting,
- code navigation,
- refactoring,
- text objects and motions,
- searching and replacing,

> and more. Tree-sitter also makes it very easy to highlight parts of a file differently if they contain code in a different language. To learn more about tree-sitter, watch [Tree-sitter - A new parsing system for programming tools - Max Brunsfield](https://www.youtube.com/watch?v=Jes3bD6P0To&feature=youtu.be&t=232).

Reviewer: mostly fine but "very easy" sounds like the the too positive language we want to avoid :]

Correspondingly, over the course of the 0.5 development cycle, there has been an explosion in the number of Lua plugins, ranging from rewrites of popular Vimscript plugins to completely new ones that would not have been possible in Vimscript -- often from contributors who were completely new to (neo)vim plugin development and were averse to learning Vimscript for that task. As a positive side effect of the long development cycle, many of these are already fully-featured and stable at the time of the 0.5 release!

Here is a small and non-representative list of Lua plugins:

- [Plenary](https://github.com/nvim-lua/plenary.nvim) -- A lua library of useful utilities for developing Neovim plugins (some of which will later be integrated into core).

> - [Packer](https://github.com/wbthomason/packer.nvim) -- A use-package like plugin manager with support for plugin dependencies, lazy-loading, and installing luarocks.

Reviewer 2: `A use-package like plugin manager ` is that a famous term ? is it the name of the emacs package manager ? in which case it may be clearer to just reference emacs.

Author: I have no idea; that I took from the previous draft. I'll just write "package manager", the rest is explained in the link.

- [Telescope](https://github.com/nvim-telescope/telescope.nvim) -- A highly extendable fuzzy finder over lists.

> - [Gitsigns](https://github.com/lewis6991/gitsigns.nvim) -- A Lua (teal) plugin for showing and interacting with changes on files a git repository (asynchronously).

Reviewer 2: `in a git repository`

Reviewer 2: you can reference teal as well https://github.com/teal-language/tl

- [Nvim-compe](https://github.com/nvim-lua/completion-nvim) -- An auto completion framework for various sources, including Neovim's builtin LSP client.
- [Nvim-dap](https://github.com/mfussenegger/nvim-dap) -- A debug adapter protocol implementation for step-through debugging of your code.
- [Colorizer](https://github.com/norcalli/nvim-colorizer.lua) -- A high-performance color highlighter for Neovim without any external dependencies.
- [Formatter](https://github.com/mhartington/formatter.nvim) -- A plugin for asynchronously executing external formatting tools on the current buffer or range.
- [Hop.nvim](https://github.com/phaazon/hop.nvim) - Easy-motion without messing with your buffer.
- [Neogit](https://github.com/TimUntersberger/neogit) - Magit but for Neovim.

Not all of these are actually written in Lua: there are many other languages (some of which are typed) that compile to Lua, e.g.,

> - [Fennel(Lisp)](https://fennel-lang.org) via [Aniseed],(https://github.com/Olical/aniseed) or [Fennel-nvim](https://github.com/jaawerth/fennel-nvim),

Reviewer 2: remove the comma after `[Aniseed]`

> - [Teal](https://github.com/teal-language/tl) (a typed dialect of Lua),

Reviewer 2: no need for parentheses

Author: I'll just remove it; the description is under the link.

- [Amulet ML](https://amulet.works),
- [MoonScript](https://github.com/leafo/moonscript) via [nvim-moonmaker](https://github.com/svermeulen/nvim-moonmaker) ,

> - [Vim9script](https://github.com/tjdevries/vim9jit) (you read that right).

Reviewer 2: @tjdevries is that something we can advertise already xD ?

Reviewer: at least it should be clear that it is quite experimental compared to the earlier items in the same list  :]

Author: @tjdevries let me know how you want this to be referenced -- if at all!

Reviewer 3: Feel free to reference it, but just say "This is an experiment to determine if it's possible for neovim to support vim9script" or something like that. It's a fun project :)

Author: the current comment is "(as a proof of concept)" -- github doesn't update these conversations...

### Lua config

> It is also possible to write user configuration in Lua: If there is an `init.lua`, it is read _instead of_ `init.vim` (these cannot coexist, and having both in your config directory will give an error), and `.lua` files in runtime directories (`plugin/`, `colorscheme/`, `after/` etc.) are sourced in addition to (after) Vimscript files. (Yes, this means that you can write your color scheme as a `.lua` file!) Note that this is entirely optional and not needed to enjoy the new functions introduced in Neovim 0.5; furthermore, not every Vimscript configuration option has a fully native Lua equivalent yet. Extending the native API to cover these as well is part of the goal for Neovim 0.6.

Reviewer 2: move the dot after the parentheses `files. (Yes,`

Author: Isn't that what is already there?

(I can remove the parentheses completely, it might be distracting here.)

For a comprehensive and up-to-date guide on using Lua for Neovim scripting and configuration, see  [Getting started using Lua in Neovim](https://github.com/nanotee/nvim-lua-guide). A good model for a Lua configuration using `init.lua` is [Defaults.nvim](https://github.com/mjlbach/defaults.nvim).

## Language Server Protocol (LSP)

The [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) (LSP) is an open, JSON-RPC-based protocol for communication between code editors and language servers, which provide programming language-specific features such as

- completion,
- hover/tooltips,
- go to definition,
- show/go to references,
- show method signatures,
- rename,
- code actions (automatic formatting, organize imports, ...),

and more.

> The idea is to separate these features into an editor-independent but language-specific server and a language-independent but editor-specific client, which communicate via the Language Server Protocol over RPC. (It should be pointed out that not every server implements every feature, and the quality of the responses can vary wildly. The "reference implementation" in VS Code also often adds non-standard features that are not covered by the LSP itself).

Reviewer 2: `over RPC. ` remove dot

Author: No, that is correct here -- there are two separate sentences in the parenthesis.

Author: But the final dot needs to be inside the parenthesis.

Neovim 0.5 provides an LSP client written (mostly) in Lua that provides a highly configurable and extensible way of accessing these features. It does not aim at competing with more feature-rich and "out-of-the-box" plugins like [CoC.nvim](https://github.com/neoclide/coc.nvim) but is meant to be tailored to your preferences (while still being useable with reasonable defaults). For an overview, see TJ DeVries' [Vimconf.live presentation](https://www.youtube.com/watch?v=C9X5VF9ASac) and his [shorter video](https://www.youtube.com/watch?v=ArwDgvYEZYk).

For many language servers, [Nvim-lspconfig](https://github.com/neovim/nvim-lspconfig) already provides the necessary configuration to set everything up easily. Alternatively, some languages also have specific LSP plugins that provide a more integrated setup, e.g., for [Java](https://github.com/mfussenegger/nvim-jdtls) and [Scala](https://github.com/ckipp01/nvim-metals).

> To learn more about LSP and how to use it in Neovim, visit [Nvim-lspconfig](https://github.com/neovim/nvim-lspconfig) and read [`:h lsp`](https://neovim.io/doc/user/lsp.html).

Reviewer 2: I would mention `:LspInfo` since it is so helpful e.g., `(:LspInfo is your friend )`

Author: That may be a bit too technical towards trouble shooting?

Reviewer 2: yeah, was unsure but though I would mention it after seeing `:h lsp`.

> Expect more work on LSP to provide more configuration options and full coverage of LSP version 3.16 (including semantic highlighting).

Reviewer 2: `Expect better coverage of the LSP spec (including, semantic highlighting) in future releases`

Author: I was referring to `lsp.config` here.

Author: And I think we are pretty good on 3.15 coverage; it's just the newest version (that got published during development) that has gaps?

## Tree-sitter

Neovim 0.5 adds **experimental** support for tree-sitter, a library that parses a piece of code into a syntax tree in an incremental and error-resilient way; this means that reparsing that code after an edit is very fast, and parsing error due to, e.g., typos remain localized and do not break parsing further down. This tree can then be efficiently queried to obtain syntax information about the code. This allows for improved and/or faster

- syntax highlighting,
- code navigation,
- refactoring,
- text objects and motions,
- searching and replacing,

and more.

> In addition, tree-sitter allows "injecting" trees for other languages, which makes it easy to support, e.g., highlighting code blocks in Markdown documents. To learn more about tree-sitter, watch [Tree-sitter - A new parsing system for programming tools - Max Brunsfield](https://www.youtube.com/watch?v=Jes3bD6P0To&feature=youtu.be&t=232).

Reviewer 2: `which makes it easy to support files that contain code in several languages`

Author: Yes, that's better (since specifically Markdown has issues :P)

- [Nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter) -- Highlighting, folds, incremental selection.
- [Playground](https://github.com/nvim-treesitter/playground) -- Utility functions to easily show the parsed tree and run queries against it.
- [Nvim-treesitter-textobjects](https://github.com/nvim-treesitter/nvim-treesitter-textobjects) -- Better text objects for vim-style motions and operators.
- [Nvim-refactor](https://github.com/nvim-treesitter/nvim-treesitter-refactor) -- Highlight definitions,  navigation, smart rename.
- [Architext](https://github.com/vigoux/architext.nvim) -- structural text editing (i.e., syntax-aware search and replace).

More information on Neovim's tree-sitter features can be found in [Nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter) and read [`:h treesitter`](https://neovim.io/doc/user/treesitter.html); you should also check out Thomas Vigouroux's [Vimconf.live presentation](https://youtu.be/EReNOGuMBmo?t=515). 

> These plugins already work well in most situations, although there are still a number of bugs and performance regressions that may occur regularly and that need to be addressed before tree-sitter support in Neovim can be declared stable. Note also that enabling tree-sitter based highlighting currently disables the internal regex-based syntax engine _completely_ (which may break other features that rely on it). Fixing these issues and improving the API (which is not stable yet) will be a major focus of the development cycle leading up to the 0.6 release.

Reviewer 2: `These plugins already work well in most situations` this has not been the case for the people around me or even myself. I fear this is going to incitate people try it only to be disappointed. We agreed that it should be presented as an experimental feature so we could tone down the excitement and start directly with `there are still a number of bugs and performance regressions`. Not to undermine the formidable effort done on treesitter but it's not as smooth as lsp yet.

Author: I completely agree; this is overselling. I would still like to start the sentence with a "nice thing" -- maybe "Although these plugins already work well in many/some situations, ..."?

Author: "Although these plugins already work well enough in some situations (file types and/or features), there are..."?

Author: (again with the parentheses ;))

Reviewer 2: `work reasonably well `is too vague,  I wouldn't mention anything positive. We dont want to encourage people who were not going to try it to try it. but maybe that's just me. Comments on social media are really positive about treesitter but I wouldn't endorse it since we strive for more :D ! but maybe that's just me.

Author: "Vague" was precisely the goal ;) And I _would_ like to mention something positive (this statement is in fact true for some languages, like Lua).

We can take a poll: thumbs up -- start with vague nice words; thumbs down -- hit them straight up with the warning stick.

Author: alternative suggestions for "work" and "well" are also welcome.

> Although these plugins already work well enough in some situations (file types and/or features), there are still a number of bugs and performance regressions that may occur regularly and that need to be addressed before tree-sitter support in Neovim can be declared stable. Note also that enabling tree-sitter based highlighting currently disables the internal regex-based syntax engine _completely_ (which may break other features that rely on it). Fixing these issues and improving the API (which is not stable yet) will be a major focus of the development cycle leading up to the 0.6 release.

Reviewer: >  Note also that enabling tree-sitter based highlighting currently disables the internal regex-based syntax engine _completely_

with the strong emphasis of "completely" it should be mentioned that is "completely" is still only per filetype.

In particular, this means that language servers can use _semantic_ information from a different file to annotate the tree for the current file: For example, a variable declared as `const` in one file can be highlighted in red if it is used in a _different_ file -- something that tree-sitter cannot do since it only has access to the latter file when highlighting.

For more details, watch TJ DeVries' [presentation on this topic](https://www.youtube.com/watch?v=c17j09vY5sw).

## Other changes

Of course, these were not the only [major changes in 0.5](https://github.com/neovim/neovim/commit/a5ac2f45ff84a688a09479f357a9909d5b914294). Here is a short summary of representative new features.

### Decorations

> There is an improved "decoration provider" API that allows setting and interacting with "extmarks" (invisible anchored text markers that move when surrounding text is edited), "virtual text" (text overlays that can now be drawn at any position on screen), and highlights. (This API is heavily leveraged by tree-sitter.)

Reviewer 2: I didn't know I had an OCD before I'd seen your parentheses :p   `(This API is heavily leveraged by tree-sitter.)`

Author: You and me, both :)

Was this an actual suggestion, though, or just copied from the text? (Because I don't see a difference...)

Reviewer: I would also remove the "quotes" around "virtual text" as that is existing terminology since quite some time :)

Author: But yeah, I can remove the parentheses with a slight reformulation.

Author: good call; no need for scare quotes.

```
See `:h vim.highlight.on_yank()` for more configuration options.

### Vim patches

Of the over 4000 commits in this release, about 1000 were patches and runtime updates ported from Vim -- nearly all of them by or with the help of the amazing [@janlazo](https://github.com/janlazo). In particular, the runtime files (syntax files, documentation, etc.) are fully synced with Vim up to [May 2021](https://github.com/neovim/neovim/pull/14424), with many later changes already included as well.

## Community

In keeping with the title of this newsletter, one of the most noticeable positive changes was the growth of the community and of new ways of interacting with it.

### Discourse

Previously, support requests and discussions were spread across Reddit, Gitter, and GitHub Discussions and were either ephemeral or hard to search for. We have now consolidated around a new [Neovim Discourse](https://neovim.discourse.group), which is a free and open source forum platform with mailing list and RSS features, in addition to a nice web interface. The Neovim Discourse is an official core project and moderated by core team members.

### Matrix

> The official chatroom for Neovim is on [Gitter](https://gitter.im/neovim/neovim). After the acquisition of Gitter by Matrix (a federated chat protocol), this room can now also be accessed from [Matrix](https://app.element.io/#/room/#neovim:matrix.org); it is also bridged to the IRC network [Libera.chat](https://libera.chat). Due to the increasing number of users, there are now additional, more specific rooms for [development of and around neovim](https://app.element.io/#/room/#nvim-dev:matrix.org), [GUIs](https://app.element.io/#/room/#neovim-guis:matrix.org), and [off-topic chat](https://app.element.io/#/room/#neovim-chat:matrix.org) -- with possibly more to come.

Reviewer 2: `-- with possibly more to come` is unnecessary.

Author: We've talked about having more rooms? But sure, I can remove it.

(The links above are to the rooms accessed through [Element](https://app.element.io), a web-based Matrix client; you can also access it through any of the many other [Matrix clients](https://matrix.org/clients/).)

### Vimconf.live

> Due to the global COVID-19 pandemic, VimConf 2020 unfortunately had to be canceled. In its palce, a virtual [Vimconf.live](https://www.vimconf.live) conference was held, with 16 speakers and over 1000 registered participants from 12 countries. If you have missed it, you can watch the lectures on the [Youtube playlist](https://www.youtube.com/playlist?list=PLcTu2VkAIIWzD2kicFNHN2c35XQCeZdsv).

Reviewer 2: pacle-> `place`

---
layout: newsletter

> title: "Newsletter #11 - The Christmas Issue"

Reviewer 4: hacker news is very strict about how submissions are titled, let's make sure Neovim is in the title.
```suggestion
title: "Neovim news #11 - The Christmas Issue"
```

Author: That'd be a "rebranding", but I like it (will insist on title case, though ;))

category: newsletter
permalink: /news/2021/07/09
---

# Neovim News #11 - The Christmas Issue

> The _real_ 0.5 was the friends we made along the way

> The long-awaited release of Neovim v0.5.0 finally happened on July 2, 2021. Due to a variety of reasons, both technical and personal (such as a certain world-wide pandemic that everyone is heartily sick of by now), this took longer than everybody had hoped for. But it's a big one! In fact, with over 4000 commits, it is so big that it broke some of the tooling for the release. So these notes will not be able to touch on each of the many changes that were made over the course of development and only focus on the most user-visible improvements, of which the biggest are

Reviewer 4: let's not even mention the pandemic, it's cliche. also don't need to make excuses about the delay, though we can acknowledge it briefly.

```suggestion
The long-awaited release of Neovim v0.5.0 finally happened on July 2, 2021. It was delayed too long, but it's worth the wait. With over 4000 commits, it is so big that it broke some of the release tooling. So these notes will not be able to touch on each of the many changes that were made over the course of development and only focus on the most user-visible improvements, of which the biggest are
```

Author: We discussed this, and it is a fact that the pandemic had a significant effect on the work (maybe not for everyone, but certainly for some). But it should be taken as given, and I can certainly tone it down.

(I wouldn't write "too long" -- it took as long as it needed to take, given the circumstances.)

Reviewer 4: sure, this also works: "It was delayed, but it's worth the wait."

Correspondingly, over the course of the 0.5 development cycle, there has been an explosion in the number of Lua plugins, ranging from rewrites of popular Vimscript plugins to completely new ones that would not have been possible in Vimscript -- often from contributors who were completely new to (neo)vim plugin development and were averse to learning Vimscript for that task. As a positive side effect of the long development cycle, many of these are already fully-featured and stable at the time of the 0.5 release!

Here is a small and non-representative list of Lua plugins:

- [Plenary](https://github.com/nvim-lua/plenary.nvim) -- A library of useful utilities for developing Neovim plugins (some of which will later be integrated into core).
- [Packer](https://github.com/wbthomason/packer.nvim) -- A package manager with support for plugin dependencies, lazy-loading, and installing luarocks.
- [Telescope](https://github.com/nvim-telescope/telescope.nvim) -- A highly extendable fuzzy finder over lists.
- [Gitsigns](https://github.com/lewis6991/gitsigns.nvim) -- A plugin for showing and interacting with changes on files in a git repository (asynchronously).

> - [Nvim-compe](https://github.com/nvim-lua/completion-nvim) -- An auto completion framework for various sources, including Neovim's builtin LSP client.

Reviewer 5: nvim-lua/completion-nvim → hrsh7th/nvim-compe ?

To learn more about LSP and how to use it in Neovim, visit [Nvim-lspconfig](https://github.com/neovim/nvim-lspconfig) and read [`:h lsp`](https://neovim.io/doc/user/lsp.html).

 Expect more work on LSP during the 0.5.x development cycle to provide improved configuration options and better coverage of the latest LSP specification (version 3.16 at the point of writing), including semantic highlighting.

## Tree-sitter

Neovim 0.5 adds **experimental** support for tree-sitter, a library that parses a piece of code into a syntax tree in an incremental and error-resilient way; this means that reparsing that code after an edit is very fast, and parsing error due to, e.g., typos remain localized and do not break parsing further down. This tree can then be efficiently queried to obtain syntax information about the code. This allows for improved and/or faster

- syntax highlighting,
- code navigation,
- refactoring,
- text objects and motions,
- searching and replacing,

and more.

> In addition, tree-sitter allows "injecting" trees for other languages, which makes it easy to support files that contain code in multiple languages. To learn more about tree-sitter, watch [Tree-sitter - A new parsing system for programming tools - Max Brunsfield](https://www.youtube.com/watch?v=Jes3bD6P0To&feature=youtu.be&t=232).

Reviewer: if the user isn't expected to know what "injecting" means perhaps it is not the best to mention upfront. maybe

> In addition, tree-sitter allows us to support files that contain code in multiple languages. such as lua blocks inside a `.vim` file or vice versa.

Author: But we don't have a vimscript parser yet 😆

Author: but point about technical term well taken; better now?

---
layout: newsletter

> title: "Newsletter #11 - Neovim 0.5"

Reviewer 2: we need something catchy: `halfway there`, "the snail edition", "the LSP apocalypse" :p

Author: I like @bfredl's quote I used as a subtitle -- should I put this in the title as well?

Author: (Right now the title is chosen to mirror that of newsletter #10)

Author: I moved the subtitle to the title -- was that something like what you have in mind?

Reviewer 2: I like the quote but it obfuscates the title

category: newsletter
permalink: /news/2021/07/5
---

# Newsletter #11: Neovim 0.5

> "The real 0.5 was the friends we made along the way."

The long-awaited release of Neovim 0.5.0 finally happened on July 2, 2021. Due to a variety of reasons, both technical and personal (such as a certain world-wide pandemic that everyone is heartily sick of by now), this took longer than everybody had hoped for. But it's a big one! In fact, with over 4000 commits, it is so big that it broke some of the tooling for the release. So these notes will not be able to touch on each of the many changes that were made over the course of development and only focus on the most user-visible improvements, of which the biggest are
1. Lua as a first-class scripting and configuration language,
2. Language server protocol (LSP),
3. Treesitter (early access).

## Lua everywhere!

Neovim 0.5 went far in making Lua a first-class scripting language for Neovim, both for plugin development and for user configuration.

> As a reminder, Lua is a small scripting language designed for embedding and is often used in, e.g., game development (Unity). Furthermore, there is a just-in-time compiler ([LuaJIT](https://github.com/LuaJIT/LuaJIT), which Neovim is built with on platforms where it is available) that can provide impressive performance on relevant tasks. In essence, Lua was chosen over other languages because it is

Reviewer 6: AFAIK, lua is not used as a scripting language in unity without a third party package. You may be thinking of Cryengine. There are of course many others.

Author: I may not have been thinking at all ;) I'll just remove the specific example; it doesn't add anything here.

1. tiny -- perfect for embedding (as opposed to a remote plugin host),
2. fast -- LuaJIT can be orders of magnitude faster than Vimscript (and non-JITed Lua),
3. simple -- small but expressive syntax (Lua 5.1) made for scripting, with Neovim itself providing a "standard library" by exposing internal functions via an API.

For more details on this choice, see Justin M Keyes' [presentation at Vim Conf 2019 ](https://www.youtube.com/watch?v=Bt-vmPC_-Ho) and TJ DeVries' [presentation at Vimconf.live](https://www.youtube.com/watch?v=IP3J56sKtn0).

Let's look at what this means for plugin authors and users.

### Lua plugins

> Neovim exposes its API natively to Lua via, e.g., `vim.api.nvim_open_win()`. It also provides methods for accessing vimscript (ex) commands and variables via, e.g., `vim.cmd(echo 'foo')` and `vim.g.syntax_on`, respectively. This allows writing plugins with the same capabilities as those written in Vimscript while exploiting the performance of Lua(JIT) at core programming language tasks like loops. It is also possible to make use of Lua's own `luarocks` plugin ecosystem.

Reviewer 2: `vim.cmd("echo 'foo'")`
